File: app/src/main/java/net/north101/android/ghplayertracker/CharacterListFragment.java
package net.north101.android.ghplayertracker;

import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.OvershootInterpolator;

import net.north101.android.ghplayertracker.data.Character;
import net.north101.android.ghplayertracker.data.CharacterList;
import net.north101.android.ghplayertracker.data.SelectableCharacter;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Background;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.UiThread;
import org.androidannotations.annotations.ViewById;
import org.json.JSONException;
import org.json.JSONObject;
import org.parceler.Parcels;

import java.io.IOException;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import jp.wasabeef.recyclerview.animators.FadeInDownAnimator;

@EFragment(R.layout.character_list_layout)
public class CharacterListFragment extends Fragment implements ActionMode.Callback {
    @ViewById(R.id.toolbar)
    Toolbar toolbar;
    @ViewById(R.id.fab)
    FloatingActionButton fab;
    @ViewById(R.id.character_list)
    RecyclerView listView;
    @ViewById(R.id.loading)
    View loadingView;

    CharacterListAdapter listAdapter;

    @AfterViews
    void afterViews() {
        ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);

        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Fragment fragment = new ClassListFragment_();
                getActivity().getSupportFragmentManager().beginTransaction()
                    .setCustomAnimations(R.anim.enter_from_right, R.anim.exit_to_left, R.anim.enter_from_left, R.anim.exit_to_right)
                    .replace(R.id.content, fragment)
                    .addToBackStack(null)
                    .commit();
            }
        });

        LinearLayoutManager listViewLayoutManager = new LinearLayoutManager(getContext());
        listView.setLayoutManager(listViewLayoutManager);

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(getContext(), listViewLayoutManager.getOrientation());
        listView.addItemDecoration(dividerItemDecoration);

        FadeInDownAnimator animator = new FadeInDownAnimator();
        animator.setInterpolator(new OvershootInterpolator());
        listView.setItemAnimator(animator);

        listAdapter = new CharacterListAdapter(new ArrayList<SelectableCharacter>());
        listAdapter.setOnClickListener(onClickListener);
        listView.setAdapter(listAdapter);

        loadCharacterList();
    }

    public ActionMode actionMode;

    @Override
    public void onPause() {
        super.onPause();
        if (actionMode != null) {
            actionMode.finish();
        }
    }

    BaseViewHolder.ClickListener<SelectableCharacter> onClickListener = new BaseViewHolder.ClickListener<SelectableCharacter>() {
        @Override
        public void onItemClick(BaseViewHolder<SelectableCharacter> holder) {
            if (actionMode != null) {
                holder.item.selected = !holder.item.selected;
                listAdapter.notifyItemChanged(holder.getAdapterPosition());
                return;
            }

            Fragment fragment = new CharacterFragment_();
            Bundle args = new Bundle();
            args.putParcelable("character", Parcels.wrap(holder.item.character));
            fragment.setArguments(args);

            getActivity().getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.enter_from_right, R.anim.exit_to_left, R.anim.enter_from_left, R.anim.exit_to_right)
                .replace(R.id.content, fragment)
                .addToBackStack(null)
                .commit();
        }

        @Override
        public boolean onItemLongClick(BaseViewHolder<SelectableCharacter> holder) {
            if (actionMode != null) {
                actionMode.finish();
                return true;
            }

            actionMode = ((AppCompatActivity) getActivity()).startSupportActionMode(CharacterListFragment.this);
            holder.item.selected = true;
            listAdapter.notifyItemChanged(holder.getAdapterPosition());

            return true;
        }
    };

    void loadCharacterList() {
        listView.setVisibility(View.INVISIBLE);
        loadingView.setVisibility(View.VISIBLE);

        loadCharacterListTask(getContext());
    }

    @Background()
    void loadCharacterListTask(Context context) {
        CharacterList characterList;
        try {
            characterList = CharacterList.load(context);
        } catch (IOException | JSONException | ParseException e) {
            e.printStackTrace();

            characterList = new CharacterList();
        }

        setCharacterList(characterList);
    }

    @UiThread(propagation = UiThread.Propagation.REUSE)
    void setCharacterList(CharacterList characterList) {
        if (this.isRemoving()) {
            return;
        }

        ArrayList<SelectableCharacter> selectableCharacterList = new ArrayList<>();
        for (Character character : characterList.values()) {
            selectableCharacterList.add(new SelectableCharacter(character, false));
        }
        Collections.sort(selectableCharacterList, new Comparator<SelectableCharacter>() {
            @Override
            public int compare(SelectableCharacter item1, SelectableCharacter item2) {
                return item2.character.getModified().compareTo(item1.character.getModified());
            }
        });

        loadingView.setVisibility(View.GONE);
        listView.setVisibility(View.VISIBLE);

        int oldSize = listAdapter.items.size();
        listAdapter.items.clear();
        listAdapter.notifyItemRangeRemoved(0, oldSize);
        listAdapter.items.addAll(selectableCharacterList);
        listAdapter.notifyItemRangeInserted(0, listAdapter.items.size());
        listView.scheduleLayoutAnimation();
    }

    @Background
    void deleteSelectedCharacters() {
        JSONObject data;
        try {
            data = CharacterList.loadJSON(getContext());
        } catch (IOException | JSONException | ParseException e) {
            e.printStackTrace();
            return;
        }

        int charactersDeleted = 0;
        for (SelectableCharacter character : listAdapter.items) {
            if (character.selected) {
                data.remove(character.character.getId().toString());
                charactersDeleted++;
            }
        }
        if (charactersDeleted > 0) {
            try {
                CharacterList.saveJSON(getContext(), data);
            } catch (IOException | JSONException e) {
                e.printStackTrace();
                Snackbar.make(getView(), "Failed to delete character(s)", Snackbar.LENGTH_SHORT).show();
                return;
            }
            setCharacterList(CharacterList.parse(getContext(), data));
        }
        Snackbar.make(getView(), "Deleted " + String.valueOf(charactersDeleted) + " character(s)", Snackbar.LENGTH_SHORT).show();

    }

    @Override
    public boolean onCreateActionMode(ActionMode actionMode, Menu menu) {
        MenuInflater inflater = actionMode.getMenuInflater();
        inflater.inflate(R.menu.character_list_action_mode, menu);
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode actionMode, Menu menu) {
        return false;
    }

    @Override
    public boolean onActionItemClicked(ActionMode actionMode, MenuItem menuItem) {
        if (menuItem.getItemId() == R.id.delete) {
            deleteSelectedCharacters();
            actionMode.finish();
            return true;
        }
        return false;
    }

    @Override
    public void onDestroyActionMode(ActionMode actionMode) {
        this.actionMode = null;
    }
}

File: app/src/main/java/net/north101/android/ghplayertracker/CharacterListFragment.kt
package net.north101.android.ghplayertracker

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.support.v7.view.ActionMode
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.mikepenz.itemanimators.SlideDownAlphaAnimator
import net.north101.android.ghplayertracker.data.Character
import net.north101.android.ghplayertracker.data.CharacterData
import net.north101.android.ghplayertracker.data.SelectableCharacter
import org.androidannotations.annotations.*
import org.json.JSONException
import java.io.IOException
import java.text.ParseException
import java.util.*
import kotlin.collections.ArrayList

@EFragment(R.layout.character_list_layout)
open class CharacterListFragment : Fragment(), ActionMode.Callback {
    @ViewById(R.id.toolbar)
    protected lateinit var toolbar: Toolbar
    @ViewById(R.id.fab)
    protected lateinit var fab: FloatingActionButton
    @ViewById(R.id.character_list)
    protected lateinit var listView: RecyclerView
    @ViewById(R.id.loading)
    protected lateinit var loadingView: View

    protected lateinit var listAdapter: CharacterListAdapter

    lateinit var classModel: ClassModel

    @JvmField
    @InstanceState
    protected var selectedCharacters = ArrayList<SelectableCharacter>()

    var actionMode: ActionMode? = null

    var onClickListener: BaseViewHolder.ClickListener<SelectableCharacter> = object : BaseViewHolder.ClickListener<SelectableCharacter>() {
        override fun onItemClick(holder: BaseViewHolder<SelectableCharacter>) {
            val actionMode = actionMode
            if (actionMode != null) {
                holder.item!!.selected = !holder.item!!.selected
                listAdapter.notifyItemChanged(holder.adapterPosition)

                if (selectedCharacters.filter { it.selected }.count() == 0) {
                    actionMode.finish()
                }
                return
            }

            val fragment = CharacterFragment_()
            val args = Bundle()
            args.putParcelable("character", holder.item!!.character.copy())
            fragment.arguments = args

            fragmentManager!!.beginTransaction()
                .replace(R.id.content, fragment)
                .addToBackStack(null)
                .commit()
        }

        override fun onItemLongClick(holder: BaseViewHolder<SelectableCharacter>): Boolean {
            if (actionMode != null) {
                actionMode!!.finish()
                return true
            }

            actionMode = (activity as AppCompatActivity).startSupportActionMode(this@CharacterListFragment)
            holder.item!!.selected = true
            listAdapter.notifyItemChanged(holder.adapterPosition)

            return true
        }
    }

    @AfterViews
    fun afterViews() {
        classModel = ViewModelProviders.of(this.activity!!).get(ClassModel::class.java)

        (activity as AppCompatActivity).setSupportActionBar(toolbar)

        fab.setOnClickListener {
            val fragment = ClassListFragment_()
            fragmentManager!!.beginTransaction()
                .replace(R.id.content, fragment)
                .addToBackStack(null)
                .commit()
        }

        val listViewLayoutManager = LinearLayoutManager(context)
        listView.layoutManager = listViewLayoutManager

        val animator = SlideDownAlphaAnimator()
        listView.itemAnimator = animator

        listAdapter = CharacterListAdapter()
        listAdapter.updateItems(selectedCharacters)
        listAdapter.setOnClickListener(onClickListener)
        listView.adapter = listAdapter

        listAdapter.updateItems(selectedCharacters)
    }

    override fun onResume() {
        super.onResume()

        view!!.post {
            classModel.characterList.observe(this, Observer {
                if (it != null) {
                    Log.d("setCharacterList", UUID.randomUUID().toString())
                    setCharacterList(it)
                }
            })
            if (classModel.characterList.value == null) {
                classModel.characterList.load()
            }
        }
    }

    override fun onPause() {
        super.onPause()
        if (actionMode != null) {
            actionMode!!.finish()
        }

        selectedCharacters = ArrayList(selectedCharacters.map { it.copy() })
        classModel.characterList.removeObservers(this)
        classModel.classList.removeObservers(this)
    }

    @UiThread(propagation = UiThread.Propagation.REUSE)
    open fun setCharacterList(characterList: List<Character>, clearSelected: Boolean = false) {
        if (this.isRemoving)
            return

        if (clearSelected)
            selectedCharacters.clear()

        loadingView.visibility = View.GONE
        listView.visibility = View.VISIBLE

        selectedCharacters = ArrayList(characterList.map {
            SelectableCharacter(it, selectedCharacters.find { selected ->
                selected.character.id == it.id
            }?.selected ?: false)
        })
        listAdapter.updateItems(selectedCharacters)
    }

    @Background
    open fun deleteCharacters(characterList: List<Character>) {
        val data = try {
            CharacterData.load(context!!)
        } catch (e: IOException) {
            e.printStackTrace()
            return
        } catch (e: JSONException) {
            e.printStackTrace()
            return
        } catch (e: ParseException) {
            e.printStackTrace()
            return
        }

        characterList.forEach { data.delete(it) }
        if (characterList.count() > 0) {
            try {
                data.save()
            } catch (e: IOException) {
                e.printStackTrace()
                Snackbar.make(activity!!.findViewById(R.id.content), "Failed to delete item(s)", Snackbar.LENGTH_SHORT).show()
                return
            } catch (e: JSONException) {
                e.printStackTrace()
                Snackbar.make(activity!!.findViewById(R.id.content), "Failed to delete item(s)", Snackbar.LENGTH_SHORT).show()
                return
            }

            view!!.post {
                classModel.characterList.load()
            }
        }
        Snackbar.make(activity!!.findViewById(R.id.content), "Deleted " + characterList.count().toString() + " item(s)", Snackbar.LENGTH_SHORT).show()
    }

    override fun onCreateActionMode(actionMode: ActionMode, menu: Menu): Boolean {
        val inflater = actionMode.menuInflater
        inflater.inflate(R.menu.character_list_action_mode, menu)
        return true
    }

    override fun onPrepareActionMode(actionMode: ActionMode, menu: Menu): Boolean {
        return false
    }

    override fun onActionItemClicked(actionMode: ActionMode, menuItem: MenuItem): Boolean {
        if (menuItem.itemId == R.id.delete) {
            deleteCharacters(selectedCharacters.filter { it.selected }.map { it.character })
            actionMode.finish()
            return true
        }
        return false
    }

    override fun onDestroyActionMode(actionMode: ActionMode) {
        selectedCharacters.forEach { it.selected = false }
        listAdapter.notifyDataSetChanged()
        this.actionMode = null
    }
}
